src/v/kafka/server/handlers/list_partition_reassignments.cc

```diff
@@ -0,0 +1,183 @@
+  cluster::partition_assignment& current_assignment,
+  std::optional<std::vector<model::broker_shard>> previous_replica_set) {
+    ongoing_partition_reassignment reassignments;
+
+    std::for_each(
+      current_assignment.replicas.cbegin(),
+      current_assignment.replicas.cend(),
+      [&reassignments,
+       &previous_replica_set](const model::broker_shard& bshard) {
+          reassignments.replicas.push_back(bshard.node_id);
+          if (previous_replica_set.has_value()) {
+              if (!in_broker_list(bshard, *previous_replica_set)) {
+                  reassignments.adding_replicas.push_back(bshard.node_id);
+              }
+          }
+      });
+
+    if (previous_replica_set.has_value()) {
+        std::for_each(
+          previous_replica_set->cbegin(),
+          previous_replica_set->cend(),
+          [&reassignments,
+           &current_assignment](const model::broker_shard& bshard) {
+              if (!in_broker_list(bshard, current_assignment.replicas)) {
+                  reassignments.removing_replicas.push_back(bshard.node_id);
+              }
+          });
+    }
+
+    return reassignments;
+}
+
+// Returns all reassignments currently in progress
+std::vector<ongoing_topic_reassignment> all_active_partition_reassignments(
+  const cluster::topic_table::updates_t& in_progress) {
+    // The in progress reassignments are in a map: ntp -> list of brokers which
+    // is not suitable to use with the ListPartitionReassignments response. So
+    // we restructure them here.
+    std::unordered_map<model::topic, ongoing_topic_reassignment>
+      reassignments_by_topic;
+
+    for (auto& [ntp, status] : in_progress) {
+        // Convert list of broker shards to node ids
+        std::vector<model::node_id> replicas;
+        std::for_each(
+          status.get_previous_replicas().begin(),
+          status.get_previous_replicas().end(),
+          [&replicas](const model::broker_shard& bshard) {
+              replicas.push_back(bshard.node_id);
+          });
+
+        if (reassignments_by_topic.contains(ntp.tp.topic)) {
+            reassignments_by_topic[ntp.tp.topic].partitions.push_back(
+              ongoing_partition_reassignment{
+                .partition_index = ntp.tp.partition, .replicas = replicas});
+        } else {
+            ongoing_topic_reassignment topic_reassignment{.name = ntp.tp.topic};
+            topic_reassignment.partitions.push_back(
+              ongoing_partition_reassignment{
+                .partition_index = ntp.tp.partition, .replicas = replicas});
+            reassignments_by_topic[ntp.tp.topic] = topic_reassignment;
+        }
+    }
+
+    std::vector<ongoing_topic_reassignment> all_active_reassignments;
+    for (auto& [_, reassignments] : reassignments_by_topic) {
+        all_active_reassignments.push_back(reassignments);
+    }
+
+    return all_active_reassignments;
+}
+
+template<>
+ss::future<response_ptr> list_partition_reassignments_handler::handle(
```

Comment:
@mmaslankaprv would be great to have you take a look at this handler.

Reply:
done